### PR TITLE
Add ctests for running GETKF in split mode and move reference files into repo 

### PR DESCRIPTION
**Pull request by SamuelDegelia-NOAA**
This PR adds ctests and yaml files for running GETKF in split observer and solver mode. Running GETKF in this form is currently the most efficient option since the observer can be run using the RoundRobin distribution (see #122). Reference files are also added for these two new ctests and are added into the repo (see below). 

There are a couple of additional changes included here: 
- All ctests now run in the same directory instead of individual directories for each test. This is done so that the GETKF solver ctest can access the hofx file output by the observer ctest. 
- The ctest reference files are moved into the repo instead of being linked in from RDAS_DATA. This will make it easier to track these small files and for other developers to add/modify them. 

Ctest output from a fresh clone: 

```
(eva) [Samuel.Degelia@hfe11 rrfs-test]$ ctest
Test project /scratch1/BMC/zrtrr/Samuel.Degelia/RDASApp_dev/RDASApp/build/rrfs-test
    Start 1: rrfs_fv3jedi_hyb_2022052619
1/8 Test #1: rrfs_fv3jedi_hyb_2022052619 ...............   Passed  146.04 sec
    Start 2: rrfs_fv3jedi_letkf_2022052619
2/8 Test #2: rrfs_fv3jedi_letkf_2022052619 .............   Passed   46.69 sec
    Start 3: rrfs_mpasjedi_2024052700_Ens3Dvar
3/8 Test #3: rrfs_mpasjedi_2024052700_Ens3Dvar .........   Passed  114.99 sec
    Start 4: rrfs_mpasjedi_2024052700_letkf
4/8 Test #4: rrfs_mpasjedi_2024052700_letkf ............   Passed   41.88 sec
    Start 5: rrfs_mpasjedi_2024052700_getkf
5/8 Test #5: rrfs_mpasjedi_2024052700_getkf ............   Passed   71.20 sec
    Start 6: rrfs_mpasjedi_2024052700_getkf_observer
6/8 Test #6: rrfs_mpasjedi_2024052700_getkf_observer ...   Passed   48.46 sec
    Start 7: rrfs_mpasjedi_2024052700_getkf_solver
7/8 Test #7: rrfs_mpasjedi_2024052700_getkf_solver .....   Passed   78.14 sec
    Start 8: rrfs_mpasjedi_2024052700_bumploc
8/8 Test #8: rrfs_mpasjedi_2024052700_bumploc ..........   Passed  146.61 sec

100% tests passed, 0 tests failed out of 8

Label Time Summary:
mpi            = 694.01 sec*proc (8 tests)
rdas-bundle    = 694.01 sec*proc (8 tests)
script         = 694.01 sec*proc (8 tests)

Total Test time (real) = 695.06 sec
```

### CODE REVIEW

File: rrfs-test/CMakeLists.txt
Comment (TingLei-NOAA) [inline rrfs-test/CMakeLists.txt]: So, multiple tests will be run in the same casedir ( defined in line 42)?

Comment (SamuelDegelia-NOAA) [reply rrfs-test/CMakeLists.txt]: Yes - all ctests are now run in the same casedir (one casedir for fv3-jedi and one casedir for mpas-jedi). I found that mpas-jedi does the same thing for their ctests when I was trying to debug the failing tests in #158.

Comment (TingLei-NOAA) [reply rrfs-test/CMakeLists.txt]: @SamuelDegelia-NOAA  I would suggest to stick to the practice that each job to be run in its' own rundir.  Otherwise, it might mess the process and make it more difficult to manage and debug if any problems occur.  Some common data could use symbolic link.

Comment (TingLei-NOAA) [reply rrfs-test/CMakeLists.txt]: Especially, when the ctests could be run simultaneously,  multiple jobs running in the same dir seems not safe to me.

Comment (TingLei-NOAA) [reply rrfs-test/CMakeLists.txt]: For clarification.  In the current case dir, are there any files or just a directory structure?  The latter is what is applied in jedi.  They have a root rundir :  .../test/ .    And all IO are in directories in this "root" rundir without interacting to each other.

File: rrfs-test/testinput/rrfs_mpasjedi_2024052700_Ens3Dvar.yaml
Comment (guoqing-noaa) [inline rrfs-test/testinput/rrfs_mpasjedi_2024052700_Ens3Dvar.yaml:128]: @SamuelDegelia-NOAA To confirm, if we don't link testoutput/ and don't do "test",  jedi_var.x will still run without an error, right?

File: rrfs-test/CMakeLists.txt
Comment (SamuelDegelia-NOAA) [reply rrfs-test/CMakeLists.txt]: This change was made because the ctest for the GETKF solver needs the output from the GETKF observer ctest. An alternative would be to add logic specific to the split GETKF ctests in `rrfs-test/CMakeLists.txt` but I prefer to make things as generic as possible. 

With the implementation in this PR, all the ctests work as expected. The output from each ctest is saved in `build/rrfs-test/Testing/Temporary/LastTest.log` which makes it easy to track any potential errors. 

> Especially, when the ctests could be run simultaneously, multiple jobs running in the same dir seems not safe to me.

I just ran all rrfs-test ctests in parallel using `ctest -j8` and they all passed. Also, is this a feature we mean to support? 

> For clarification. In the current case dir, are there any files or just a directory structure? The latter is what is applied in jedi. They have a root rundir : .../test/ . And all IO are in directories in this "root" rundir without interacting to each other.

The current casedir links in the `data` input directory but there are also a few fix files copied into the casedir (e.g., `namelist.atmosphere`, `geovars.yaml`). The ctests read these files but they do not modify anything. The ctests output various files but I modified some of the yamls so that the output files are specific to those tests (e.g., `ana_getkf.2024-05-27_00.00.00.nc`).

File: rrfs-test/testinput/rrfs_mpasjedi_2024052700_Ens3Dvar.yaml
Comment (SamuelDegelia-NOAA) [reply rrfs-test/testinput/rrfs_mpasjedi_2024052700_Ens3Dvar.yaml:128]: If we do not link testoutput in, JEDI will complete (e.g., outputting all analysis or bumploc files) but it will throw an error at the end for the reference mismatch error. 

I would like to keep these yaml files in `rrfs-test/testinput` as inputs designed for the ctests. If you are worried about using these yamls for separate experiments, then we could either 1) add a line in `setup_experiment.sh` to copy `testoutput` into the run directory, or 2) use different yaml files.

File: rrfs-test/CMakeLists.txt
Comment (SamuelDegelia-NOAA) [reply rrfs-test/CMakeLists.txt]: I don't mind setting it back to the original practice @TingLei-NOAA - it is an easy change. But just giving my two cents here.

Comment (TingLei-NOAA) [reply rrfs-test/CMakeLists.txt]: @SamuelDegelia-NOAA  I would suggest either to have different rundir for different runs/ctests or as JEDI does, have a "super" rundir without any files except for directories to keep a clean rundir. Otherwise, especially when different developers are working on this part, the problem might occur .

Comment (SamuelDegelia-NOAA) [reply rrfs-test/CMakeLists.txt]: Is the concern that each test could be outputting in the same directory simultaneously? Because they do output into the main casedir, but they are outputting different files and those should not overlap. Or is the concern about reading the input files at the same time? Because the previous method of linking in the same input `data/` directory for each test would have the same issue. Just trying to understand your concern a little more.

File: rrfs-test/testinput/rrfs_mpasjedi_2024052700_Ens3Dvar.yaml
Comment (guoqing-noaa) [reply rrfs-test/testinput/rrfs_mpasjedi_2024052700_Ens3Dvar.yaml:128]: > If we do not link testoutput in, JEDI will complete (e.g., outputting all analysis or bumploc files) but it will throw an error at the end for the reference mismatch error.
> 

If it throws an error even if we don't do ctest, then we need to address this problem in this PR: 

1. modify `link_mpasjedi_expr.sh` to link `testoutput` instead of 'ref` to under data/
Or
2. put the ref files still under `fix/expr_data/mpas_2024052700/ref` but without links. We can track them directly. 
Thanks!

File: rrfs-test/CMakeLists.txt
Comment (guoqing-noaa) [reply rrfs-test/CMakeLists.txt]: I think as long as no files (including logs) are overwritten, it should be fine to use the same directory. As @SamuelDegelia-NOAA mentioned. the JCSDA mpasjedi tests have already used this strategy.

File: rrfs-test/testinput/rrfs_mpasjedi_2024052700_Ens3Dvar.yaml
Comment (SamuelDegelia-NOAA) [reply rrfs-test/testinput/rrfs_mpasjedi_2024052700_Ens3Dvar.yaml:128]: I prefer solution #&#x2060;1 to keep all the test reference data in one location and to be consistent with other JCSDA repos. I will update the PR to do this.

Comment (guoqing-noaa) [reply rrfs-test/testinput/rrfs_mpasjedi_2024052700_Ens3Dvar.yaml:128]: Works for me. Just make sure we can run standalone mpasjedi tests without an error. Thanks

File: rrfs-test/CMakeLists.txt
Comment (SamuelDegelia-NOAA) [reply rrfs-test/CMakeLists.txt]: Update: I tested running `ctest -j8` a few more times and all tests continued to pass except for one failure of `rrfs_mpasjedi_2024052700_getkf_solver`. This test failed because the observer test needs to run first.

Comment (guoqing-noaa) [reply rrfs-test/CMakeLists.txt]: @SamuelDegelia-NOAA Looks good! 
I am always curious whether we can set dependencies when running ctests in parallel?

Comment (SamuelDegelia-NOAA) [reply rrfs-test/CMakeLists.txt]: @guoqing-noaa I believe this is an option. Here is an example that I found on [this page](https://jointcenterforsatellitedataassimilation-jedi-docs.readthedocs-hosted.com/en/latest/inside/testing/unit_testing.html) of the JEDI documentation: 

```
ecbuild_add_test( TARGET       test_fv3jedi_forecast_fv3-gfs_compare
                  TYPE         SCRIPT
                  COMMAND      ${CMAKE_BINARY_DIR}/bin/compare.py
                  ARGS         testoutput/forecast_fv3-gfs.run testoutput/forecast_fv3-gfs.ref 1.0e-3 0
                  TEST_DEPENDS test_fv3jedi_forecast_fv3-gfs )
```

I can implement this for the getkf_solver test so that all ctests can be run in parallel.

Comment (SamuelDegelia-NOAA) [reply rrfs-test/CMakeLists.txt]: Yes this functionality works. You can see here that it waits to start the solver ctest until after the observer test has passed: 

```
(eva) [Samuel.Degelia@hfe08 rrfs-test]$ ctest -j8
Test project /scratch1/BMC/zrtrr/Samuel.Degelia/RDASApp_dev/RDASApp/build/rrfs-test
    Start 1: rrfs_fv3jedi_hyb_2022052619
    Start 2: rrfs_fv3jedi_letkf_2022052619
    Start 3: rrfs_mpasjedi_2024052700_Ens3Dvar
    Start 4: rrfs_mpasjedi_2024052700_letkf
    Start 5: rrfs_mpasjedi_2024052700_getkf
    Start 6: rrfs_mpasjedi_2024052700_getkf_observer
    Start 8: rrfs_mpasjedi_2024052700_bumploc
1/8 Test #2: rrfs_fv3jedi_letkf_2022052619 .............   Passed   46.69 sec
2/8 Test #1: rrfs_fv3jedi_hyb_2022052619 ...............   Passed  109.14 sec
3/8 Test #3: rrfs_mpasjedi_2024052700_Ens3Dvar .........   Passed  221.35 sec
4/8 Test #4: rrfs_mpasjedi_2024052700_letkf ............   Passed  268.31 sec
5/8 Test #6: rrfs_mpasjedi_2024052700_getkf_observer ...   Passed  299.60 sec
    Start 7: rrfs_mpasjedi_2024052700_getkf_solver
6/8 Test #5: rrfs_mpasjedi_2024052700_getkf ............   Passed  312.33 sec
7/8 Test #7: rrfs_mpasjedi_2024052700_getkf_solver .....   Passed   55.73 sec
8/8 Test #8: rrfs_mpasjedi_2024052700_bumploc ..........   Passed  415.90 sec

100% tests passed, 0 tests failed out of 8

Label Time Summary:
mpi            = 1729.04 sec*proc (8 tests)
rdas-bundle    = 1729.04 sec*proc (8 tests)
script         = 1729.04 sec*proc (8 tests)

Total Test time (real) = 415.92 sec
```

File: rrfs-test/testinput/rrfs_mpasjedi_2024052700_Ens3Dvar.yaml
Comment (SamuelDegelia-NOAA) [reply rrfs-test/testinput/rrfs_mpasjedi_2024052700_Ens3Dvar.yaml:128]: Done. Just updated `link_mpasjedi_expr.sh` to link testouput into `expr/mpas_2024052700`. I ran `run_jedi.sh` in there and it passes without error.

**Review by guoqing-noaa — DISMISSED** [review]
LGTM. But I will do a fresh test on Jet and update the results in about 40 minutes.

**Comment by SamuelDegelia-NOAA** [comment]
> LGTM. But I will do a fresh test on Jet and update the results in about 40 minutes.

Could you do a test with `ctest -j8` too since that has been a point of discussion in this PR? Would like to confirm that it works on other machines too. Thanks!

**Comment by guoqing-noaa** [comment]
> > LGTM. But I will do a fresh test on Jet and update the results in about 40 minutes.
> 
> Could you do a test with `ctest -j8` too since that has been a point of discussion in this PR? Would like to confirm that it works on other machines too. Thanks!

Yes, I think it is a good idea to run `ctest -j8` by default for any future rdas_build_tests.

**Comment by guoqing-noaa** [comment]
Well, @SamuelDegelia-NOAA could you modify the last line of `ush/run_rrfs_tests.sh` to `ctest -j8` in this PR? Thanks!

**Comment by SamuelDegelia-NOAA** [comment]
> Well, @SamuelDegelia-NOAA could you modify the last line of `ush/run_rrfs_tests.sh` to `ctest -j8` in this PR? Thanks!

Done!

**Comment by guoqing-noaa** [comment]
all rrfs tests passed when running from a fresh clone/build on Jet:
`/mnt/lfs5/BMC/wrfruc/gge/tmp/rdas_build_test/RDASApp_SamuelDegelia-NOAA_feature_split_getkf_test`

The ctest dependency worked as expected.

**Review by guoqing-noaa — DISMISSED** [review]

**Comment by SamuelDegelia-NOAA** [comment]
We had some internal discussion today at EMC and would like to keep individual run directories for each ctest. I will update this PR today to account for this change. It should not change much but we will need to retest.

**Comment by SamuelDegelia-NOAA** [comment]
I updated the PR so now each ctest runs in its own unique directory again. The GETKF solver ctest now uses relative paths to get the hofx file output by the observer ctest: 

`obsfile: ../rundir-rrfs_mpasjedi_2024052700_getkf_observer/adpupa_hofx.nc4`

Please have another look @TingLei-NOAA to see if this works for you. 

Ctest output from a fresh clone: 

```
(eva) [Samuel.Degelia@hfe01 rrfs-test]$ ctest -j8
Test project /scratch1/BMC/zrtrr/Samuel.Degelia/RDASApp_dev/RDASApp/build/rrfs-test
    Start 6: rrfs_mpasjedi_2024052700_getkf_observer
    Start 3: rrfs_mpasjedi_2024052700_Ens3Dvar
    Start 8: rrfs_mpasjedi_2024052700_bumploc
    Start 1: rrfs_fv3jedi_hyb_2022052619
    Start 4: rrfs_mpasjedi_2024052700_letkf
    Start 5: rrfs_mpasjedi_2024052700_getkf
    Start 2: rrfs_fv3jedi_letkf_2022052619
1/8 Test #2: rrfs_fv3jedi_letkf_2022052619 .............   Passed   41.68 sec
2/8 Test #5: rrfs_mpasjedi_2024052700_getkf ............   Passed   46.51 sec
3/8 Test #4: rrfs_mpasjedi_2024052700_letkf ............   Passed   85.65 sec
4/8 Test #1: rrfs_fv3jedi_hyb_2022052619 ...............   Passed  105.76 sec
5/8 Test #6: rrfs_mpasjedi_2024052700_getkf_observer ...   Passed  121.45 sec
    Start 7: rrfs_mpasjedi_2024052700_getkf_solver
6/8 Test #3: rrfs_mpasjedi_2024052700_Ens3Dvar .........   Passed  163.80 sec
7/8 Test #8: rrfs_mpasjedi_2024052700_bumploc ..........   Passed  204.52 sec
8/8 Test #7: rrfs_mpasjedi_2024052700_getkf_solver .....   Passed   89.60 sec

100% tests passed, 0 tests failed out of 8

Label Time Summary:
mpi            = 858.97 sec*proc (8 tests)
rdas-bundle    = 858.97 sec*proc (8 tests)
script         = 858.97 sec*proc (8 tests)

Total Test time (real) = 211.05 sec
```

**Comment by ShunLiu-NOAA** [comment]
@TingLei-NOAA and @guoqing-noaa Please review this PR again.

File: rrfs-test/CMakeLists.txt
Comment (guoqing-noaa) [inline rrfs-test/CMakeLists.txt]: It looks like the indentation of line 108 is not aligned with line 101

Comment (SamuelDegelia-NOAA) [reply rrfs-test/CMakeLists.txt]: Thanks for catching this. Should be resolved now.

File: rrfs-test/scripts/link_mpasjedi_expr.sh
Comment (guoqing-noaa) [inline rrfs-test/scripts/link_mpasjedi_expr.sh:1]: Line 50: remove "ref"
remove line 56.

Comment (guoqing-noaa) [reply rrfs-test/scripts/link_mpasjedi_expr.sh:1]: Otherwise, the build.sh will fail as:
```
Linking in test data for MPAS-JEDI case
expdir is at: /work/noaa/wrfruc/gge/rdas_build_test_orion/RDASApp_SamuelDegelia-NOAA_feature_split_getkf_test/expr/mpas_2024052700
ln: ref/: cannot overwrite directory
```

This demonstrates why we always want to do a fresh clone, build, and test before merging any PRs (which change codes/scripts).

Comment (SamuelDegelia-NOAA) [reply rrfs-test/scripts/link_mpasjedi_expr.sh:1]: This did not throw an error for me on Hera. The link script created the `ref` directory which just looked like this inside: 

```
(eva) [Samuel.Degelia@hfe07 RDASApp]$ ls expr/mpas_2024052700/data/ref
'*'

```
As mentioned in my comment [here](https://github.com/NOAA-EMC/RDASApp/pull/175#issuecomment-2362102589), I tested the most recent version (without the whitespace change) using a fresh clone, build, and ctest. But thank you for pointing out that this does throw an error on Orion/Hercules. So it seems like linking is handled differently on different systems. We should thus probably test on multiple machines if we are making changes to links or building. 

Anyways, I just pushed a new commit to remove the lines you requested.

Comment (guoqing-noaa) [reply rrfs-test/scripts/link_mpasjedi_expr.sh:1]: @SamuelDegelia-NOAA Thanks for making the changes!
Yes, you made a good point. We would like to test PRs on different platforms.

**Comment by guoqing-noaa** [comment]
All rrfs tests passed on Hercules

**Review by guoqing-noaa — DISMISSED** [review]

**Comment by SamuelDegelia-NOAA** [comment]
Thank you @guoqing-noaa for testing!

**Comment by TingLei-NOAA** [comment]
On hera ,  a fresh/clean cloning and building succeeded.  All rrfs tests passed.
Some mpas-jedi tests passed I think because different crtm versions are being used. 
So, all those tests failing for non-identical results could rather safely be ignored. But some tests failed for missing files. I would come back after a closer look.

**Comment by SamuelDegelia-NOAA** [comment]
Just to confirm @TingLei-NOAA, those failing tests are not for rrfs-test? If so then is this PR good to go pending your approval?

**Comment by TingLei-NOAA** [comment]
It is found the previously reported failures of fv3-jedi  concerning me for missing files were generated for my running ctests in parallel mode while the independency between some fv3-jedi ctests are not labelled accordingly. 
After re-running ctests in the serial mode and all fv3-jedi ctets passed.

**Review by TingLei-NOAA — APPROVED** [review]

**Comment by ShunLiu-NOAA** [comment]
@SamuelDegelia-NOAA @TingLei-NOAA, @guoqing-noaa and @rrfsbot, thank you for your effort. The PR is good to merge now.

**Review by ShunLiu-NOAA — APPROVED** [review]

**Comment by SamuelDegelia-NOAA** [comment]
Thanks @TingLei-NOAA for investigating!

**Pull request merged**
